File: app/src/test/java/com/mmwtl/atlasmediawidget/MediaBridgeContractTest.java
package com.mmwtl.atlasmediawidget;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.junit.Assert.assertEquals;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 30)
public class MediaBridgeContractTest {
    @Test public void protocolV1ConstantsMatchAtlasMediaApi() {
        assertEquals("com.mmwtl.atlasmediaapi.media.BIND", MediaBridgeContract.SERVICE_ACTION);
        assertEquals("com.mmwtl.atlasmediawidget",
                MediaBridgeContract.SERVICE_PACKAGE);
        assertEquals("com.mmwtl.atlasmediaapi.media.bridge.MediaBridgeService",
                MediaBridgeContract.SERVICE_CLASS);
        assertEquals(1, MediaBridgeContract.VERSION);
        assertEquals(1, MediaBridgeContract.REGISTER);
        assertEquals(4, MediaBridgeContract.COMMAND);
        assertEquals(5, MediaBridgeContract.GET_RADIO_STATIONS);
        assertEquals(100, MediaBridgeContract.REGISTERED);
        assertEquals(103, MediaBridgeContract.ERROR);
        assertEquals(104, MediaBridgeContract.RADIO_STATIONS);
        assertEquals(0x40L, MediaBridgeContract.CAP_SET_SOURCE);
        assertEquals(0x80L, MediaBridgeContract.CAP_TUNE_RADIO);
        assertEquals("radioFavoriteStations",
                MediaBridgeContract.K_RADIO_FAVORITE_STATIONS);
        assertEquals("uiScaleTenths", MediaBridgeContract.K_UI_SCALE_TENTHS);
        assertEquals(6, MediaBridgeContract.GET_SETTINGS);
        assertEquals(7, MediaBridgeContract.UPDATE_SETTINGS);
        assertEquals(8, MediaBridgeContract.EXPORT_MEDIA_BACKUP);
        assertEquals(9, MediaBridgeContract.PREPARE_MEDIA_IMPORT);
        assertEquals(10, MediaBridgeContract.COMMIT_MEDIA_IMPORT);
        assertEquals(11, MediaBridgeContract.GET_IMPORT_STATUS);
        assertEquals(12, MediaBridgeContract.ABORT_MEDIA_IMPORT);
        assertEquals(13, MediaBridgeContract.RESTORE_DEFAULT_CATALOG);
        assertEquals(105, MediaBridgeContract.SETTINGS);
        assertEquals(106, MediaBridgeContract.SETTINGS_UPDATED);
        assertEquals(107, MediaBridgeContract.MEDIA_BACKUP_EXPORTED);
        assertEquals(108, MediaBridgeContract.MEDIA_IMPORT_PREPARED);
        assertEquals(109, MediaBridgeContract.MEDIA_IMPORT_COMMITTED);
        assertEquals(110, MediaBridgeContract.MEDIA_IMPORT_STATUS);
        assertEquals(111, MediaBridgeContract.MEDIA_IMPORT_ABORTED);
        assertEquals(112, MediaBridgeContract.DEFAULT_CATALOG_RESTORED);
        assertEquals(9, MediaBridgeContract.STATUS_VALIDATION_ERROR);
        assertEquals(10, MediaBridgeContract.STATUS_CONFLICT);
        assertEquals(11, MediaBridgeContract.STATUS_IO_ERROR);
        assertEquals("settingsRevision", MediaBridgeContract.K_SETTINGS_REVISION);
        assertEquals("defaultAudioSource", MediaBridgeContract.K_DEFAULT_AUDIO_SOURCE);
    }

    @Test public void radioTransferConstantsMatchRuntimeContract() {
        assertEquals(com.mmwtl.atlasmediaapi.media.bridge.MediaBridgeContract.ClientMessage.EXPORT_RADIO_CATALOG,
                MediaBridgeContract.EXPORT_RADIO_CATALOG);
        assertEquals(com.mmwtl.atlasmediaapi.media.bridge.MediaBridgeContract.ClientMessage.IMPORT_RADIO_CATALOG,
                MediaBridgeContract.IMPORT_RADIO_CATALOG);
        assertEquals(com.mmwtl.atlasmediaapi.media.bridge.MediaBridgeContract.ServerMessage.RADIO_CATALOG_EXPORTED,
                MediaBridgeContract.RADIO_CATALOG_EXPORTED);
        assertEquals(com.mmwtl.atlasmediaapi.media.bridge.MediaBridgeContract.ServerMessage.RADIO_CATALOG_IMPORTED,
                MediaBridgeContract.RADIO_CATALOG_IMPORTED);
        assertEquals(com.mmwtl.atlasmediaapi.media.bridge.MediaBridgeContract.Key.FILE_DESCRIPTOR,
                MediaBridgeContract.K_FILE_DESCRIPTOR);
    }

    @Test public void mediaSettingsSnapshotRoundTrip() {
        MediaSettingsSnapshot original = new MediaSettingsSnapshot(
                123L, "RADIO", 4, true, true, false, "com.test.player",
                true, true, false, true, true, true, true, 2000L, 200L,
                "CUSTOM", 15, "Test Catalog", 18);
        MediaSettingsSnapshot restored = MediaSettingsSnapshot.fromBundle(original.toBundle());
        assertEquals(original.revision, restored.revision);
        assertEquals(original.defaultAudioSource, restored.defaultAudioSource);
        assertEquals(original.defaultAudioSourceDelaySec, restored.defaultAudioSourceDelaySec);
        assertEquals(original.defaultAudioSourceAutoplayOnStartup, restored.defaultAudioSourceAutoplayOnStartup);
        assertEquals(original.autoSwitchToDefaultOnSourceLost, restored.autoSwitchToDefaultOnSourceLost);
        assertEquals(original.autoSwitchToDefaultAutoplayOnSourceLost, restored.autoSwitchToDefaultAutoplayOnSourceLost);
        assertEquals(original.defaultMediaPackage, restored.defaultMediaPackage);
        assertEquals(original.minimizeOnlinePlayerAfterAutostart,
                restored.minimizeOnlinePlayerAfterAutostart);
        assertEquals(original.switchToOnlineBeforeSessionPlay, restored.switchToOnlineBeforeSessionPlay);
        assertEquals(original.radioWidgetBroadcastEnabled, restored.radioWidgetBroadcastEnabled);
        assertEquals(original.clusterCoversEnabled, restored.clusterCoversEnabled);
        assertEquals(original.clusterRadioFacadeEnabled, restored.clusterRadioFacadeEnabled);
        assertEquals(original.clusterOnlineEnabled, restored.clusterOnlineEnabled);
        assertEquals(original.clusterOnlineProgressEnabled, restored.clusterOnlineProgressEnabled);
        assertEquals(original.clusterWatchdogIntervalMs, restored.clusterWatchdogIntervalMs);
        assertEquals(original.clusterReassertBurstIntervalMs, restored.clusterReassertBurstIntervalMs);
        assertEquals(original.catalogType, restored.catalogType);
        assertEquals(original.catalogStationCount, restored.catalogStationCount);
        assertEquals(original.catalogDescription, restored.catalogDescription);
        assertEquals(original.uiScaleTenths, restored.uiScaleTenths);
    }
}

File: media-core/src/test/kotlin/com/mmwtl/atlasmediaapi/media/bridge/MediaBridgeContractTest.kt
package com.mmwtl.atlasmediaapi.media.bridge

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class MediaBridgeContractTest {
    @Test
    fun `only the advertised protocol version is accepted`() {
        assertEquals(
            MediaBridgeContract.Status.UNSUPPORTED_VERSION,
            MediaBridgeContract.negotiate(MediaBridgeContract.MIN_PROTOCOL_VERSION - 1),
        )
        assertEquals(
            MediaBridgeContract.Status.OK,
            MediaBridgeContract.negotiate(MediaBridgeContract.PROTOCOL_VERSION),
        )
        assertEquals(
            MediaBridgeContract.Status.UNSUPPORTED_VERSION,
            MediaBridgeContract.negotiate(MediaBridgeContract.MAX_PROTOCOL_VERSION + 1),
        )
    }

    @Test
    fun `contract constants match expected values`() {
        assertEquals("com.mmwtl.atlasmediaapi.media.BIND", MediaBridgeContract.SERVICE_ACTION)
        assertEquals("com.mmwtl.atlasmediaapi", MediaBridgeContract.SERVICE_PACKAGE)
        assertEquals("com.mmwtl.atlasmediaapi.media.bridge.MediaBridgeService", MediaBridgeContract.SERVICE_CLASS)
        assertEquals(1, MediaBridgeContract.PROTOCOL_VERSION)
        assertEquals(1, MediaBridgeContract.ClientMessage.REGISTER)
        assertEquals(2, MediaBridgeContract.ClientMessage.UNREGISTER)
        assertEquals(3, MediaBridgeContract.ClientMessage.GET_SNAPSHOT)
        assertEquals(4, MediaBridgeContract.ClientMessage.COMMAND)
        assertEquals(5, MediaBridgeContract.ClientMessage.GET_RADIO_STATIONS)
        assertEquals(100, MediaBridgeContract.ServerMessage.REGISTERED)
        assertEquals(101, MediaBridgeContract.ServerMessage.SNAPSHOT)
        assertEquals(102, MediaBridgeContract.ServerMessage.COMMAND_RESULT)
        assertEquals(103, MediaBridgeContract.ServerMessage.ERROR)
        assertEquals(104, MediaBridgeContract.ServerMessage.RADIO_STATIONS)
        assertEquals(6, MediaBridgeContract.ClientMessage.GET_SETTINGS)
        assertEquals(7, MediaBridgeContract.ClientMessage.UPDATE_SETTINGS)
        assertEquals(8, MediaBridgeContract.ClientMessage.EXPORT_MEDIA_BACKUP)
        assertEquals(9, MediaBridgeContract.ClientMessage.PREPARE_MEDIA_IMPORT)
        assertEquals(10, MediaBridgeContract.ClientMessage.COMMIT_MEDIA_IMPORT)
        assertEquals(11, MediaBridgeContract.ClientMessage.GET_IMPORT_STATUS)
        assertEquals(12, MediaBridgeContract.ClientMessage.ABORT_MEDIA_IMPORT)
        assertEquals(13, MediaBridgeContract.ClientMessage.RESTORE_DEFAULT_CATALOG)
        assertEquals(14, MediaBridgeContract.ClientMessage.EXPORT_RADIO_CATALOG)
        assertEquals(15, MediaBridgeContract.ClientMessage.IMPORT_RADIO_CATALOG)
        assertEquals(105, MediaBridgeContract.ServerMessage.SETTINGS)
        assertEquals(106, MediaBridgeContract.ServerMessage.SETTINGS_UPDATED)
        assertEquals(107, MediaBridgeContract.ServerMessage.MEDIA_BACKUP_EXPORTED)
        assertEquals(108, MediaBridgeContract.ServerMessage.MEDIA_IMPORT_PREPARED)
        assertEquals(109, MediaBridgeContract.ServerMessage.MEDIA_IMPORT_COMMITTED)
        assertEquals(110, MediaBridgeContract.ServerMessage.MEDIA_IMPORT_STATUS)
        assertEquals(111, MediaBridgeContract.ServerMessage.MEDIA_IMPORT_ABORTED)
        assertEquals(112, MediaBridgeContract.ServerMessage.DEFAULT_CATALOG_RESTORED)
        assertEquals(114, MediaBridgeContract.ServerMessage.RADIO_CATALOG_EXPORTED)
        assertEquals(115, MediaBridgeContract.ServerMessage.RADIO_CATALOG_IMPORTED)
        assertEquals(9, MediaBridgeContract.Status.VALIDATION_ERROR)
        assertEquals(10, MediaBridgeContract.Status.CONFLICT)
        assertEquals(11, MediaBridgeContract.Status.IO_ERROR)
        assertEquals("uiScaleTenths", MediaBridgeContract.Key.UI_SCALE_TENTHS)
        assertEquals("settingsRevision", MediaBridgeContract.Key.SETTINGS_REVISION)
        assertEquals("defaultAudioSource", MediaBridgeContract.Key.DEFAULT_AUDIO_SOURCE)
    }

    @Test
    fun `media settings snapshot bundle round trip preserves all fields`() {
        val snapshot = MediaSettingsSnapshot(
            revision = 42L,
            defaultAudioSource = "RADIO",
            defaultAudioSourceDelaySec = 5,
            defaultAudioSourceAutoplayOnStartup = false,
            autoSwitchToDefaultOnSourceLost = true,
            autoSwitchToDefaultAutoplayOnSourceLost = false,
            defaultMediaPackage = "ru.yandex.music",
            switchToOnlineBeforeSessionPlay = true,
            radioWidgetBroadcastEnabled = false,
            clusterCoversEnabled = false,
            clusterOnlineEnabled = true,
            clusterOnlineProgressEnabled = true,
            clusterWatchdogIntervalMs = 2500L,
            clusterReassertBurstIntervalMs = 200L,
            catalogType = "CUSTOM",
            catalogStationCount = 10,
            catalogDescription = "Custom Catalog",
            uiScaleTenths = 18,
        )
        val bundle = snapshot.toBundle()
        val restored = bundle.toMediaSettingsSnapshot()
        assertEquals(snapshot, restored)
    }

    @Test
    fun `default media source capabilities include playback controls for CPAA, BT, USB, Radio`() {
        val cpaaCaps = BridgeAudioSource.CPAA.defaultCapabilities()
        val expected = MediaCapabilities.BASIC_PLAYBACK or
                MediaCapabilities.TRACK_NAVIGATION or
                MediaCapabilities.SET_SOURCE
        assertEquals(expected, cpaaCaps)
        assertEquals(expected, BridgeAudioSource.BT.defaultCapabilities())
        assertEquals(expected, BridgeAudioSource.USB.defaultCapabilities())
        assertEquals(
            expected or MediaCapabilities.TUNE_RADIO,
            BridgeAudioSource.RADIO.defaultCapabilities(),
        )
    }
}
